databallpy/features/covered_distance.py

```diff
@@ -0,0 +1,473 @@
+    
+    return [(min(intervals[i], intervals[i+1]), max(intervals[i], intervals[i+1])) for i in range(0, len(intervals), 2)]
+
+def get_covered_distance(
+    tracking_data: pd.DataFrame,
+    player_ids: list[str],
+    framerate: int,
+    velocity_intervals: tuple[float, ...] | None = None,
+    acceleration_intervals: tuple[float, ...] | None = None
+    ) -> dict:
+    """Function that calculates the distance covered in specified velocity and acceleration intervals
+
+    Args:
+        tracking_data:      tracking data
+        player_ids:         list containing players. Example = ['home_8', 'away_3', 'home_16']
+        framerate:          sample frequency in Hertz
+        velocity_intervals: tuple that contains the velocity interval(s). For example [0 3] & [4 8]: (0,3,8,4)
+        acc_intervals:      tuple that contains the acceleration interval(s). For example [-3 0] & [3 5]: (-3,0,3,5)      
+
+    Returns:
+        dict: a dictionary with for every player the total distance covered and optionally the 
+        distance covered within the given velocity and/or acceleration threshold(s)
+    
+    Notes: 
+        The function requires the velocity for every player calculated with the add_velocity function. 
+        The acceleration for every player depends on the presence of acceleration intervals in the input
+    """
+
+    try:
+        # Check input types
+        # tracking_data
+        if not isinstance(tracking_data, pd.DataFrame):
+            raise TypeError(f"tracking data must be a pandas DataFrame, not a {type(tracking_data).__name__}")
+        
+        # player_ids
+        if not isinstance(player_ids, list):
+            raise TypeError(f"player_ids must be a list, not a {type(player_ids).__name__}")
+
+        if not all(isinstance(player, str) for player in player_ids):
+            raise TypeError("All elements in player_ids must be strings")
+        
+        # framerate
+        if not isinstance(framerate, int):
+            raise TypeError(f"framerate must be a int, not a {type(framerate).__name__}")
+
+        # check velocity
+        for player_id in player_ids:
+            if (
+                player_id + "_vx" not in tracking_data.columns
+                or player_id + "_vy" not in tracking_data.columns
+                or player_id + "_velocity" not in tracking_data.columns
+            ):
+                raise ValueError(
+                    f"Velocity was not found for {player_id} in the DataFrame. "
+                    " Please calculate velocity first using add_velocity() function."
+                )
+
+        # initialize dictionary covered distance
+        players = {player_id: {'total_distance': 0} for player_id in player_ids}
+
+        # Calculate total distance
+        tracking_data_velocity = pd.concat([tracking_data[player_id + '_velocity'] for player_id in player_ids], axis=1)
+        total_distance = tracking_data_velocity.apply(lambda player: np.sum(player / framerate))
```

Comment:
check with comment above. 

Also, I think it would be good to have the covered distance per frame per player in this dataframe, instead of the total distance:

`distance_per_frame = tracking_data_velocity.diff(axis=0).fillna(0)`

Than  use masks to create a subset based on velocity and acceleration intervals
`vel_mask = tracking_data_velocity.between(start_interval, end_interval)`
`distance_covered_in_interval = distance_per_frame[vel_mask].sum()`  

(I did not run this code but should be something like this. Than you have for each player in each interval the distance covered, without needing to recalculate the distance every time)

Reply:
I don't really understand the .diff function in this case. I divided the velocity by the framerate to get the distance_per_frame. For the rest, good suggestions!